src/app/games/diaballik/DiaballikMoveGenerator.ts

```diff
@@ -5,61 +5,146 @@ import { DiaballikNode, DiaballikRules } from './DiaballikRules';
+        if (this.subMoves.length > 1) {
+            second = MGPOptional.of(this.subMoves[1]);
+        }
+        let third: MGPOptional<DiaballikSubMove> = MGPOptional.empty();
+        if (this.subMoves.length > 2) {
+            third = MGPOptional.of(this.subMoves[2]);
+        }
+        const move: DiaballikMove = new DiaballikMove(first, second, third);
+        Utils.assert(DiaballikRules.get().isLegal(move, state).isSuccess(),
+                     'DiaballikMoveGenerator should only generate legal moves');
+        return move;
+    }
+}
 
 export class DiaballikMoveGenerator extends MoveGenerator<DiaballikMove, DiaballikState> {
 
     public getListMoves(node: DiaballikNode): DiaballikMove[] {
-        const moves: DiaballikMove[] = [];
-        // We only generate moves that have 3 submoves.
-        // The reasoning is that it is super rare to not want 3 submoves.
-        const state: DiaballikState = node.gameState;
-        for (const first of this.getListSubMoves(state)) {
-            const firstLegality: MGPFallible<DiaballikState> =
-                DiaballikRules.get().isLegalSubMove(state, first);
-            if (firstLegality.isSuccess()) {
-                const stateAfterFirst: DiaballikState = firstLegality.get();
-                for (const second of this.getListSubMoves(stateAfterFirst)) {
-                    const secondLegality: MGPFallible<DiaballikState> =
-                        DiaballikRules.get().isLegalSubMove(state, second);
-                    if (secondLegality.isSuccess()) {
-                        const stateAfterSecond: DiaballikState = secondLegality.get();
-                        for (const third of this.getListSubMoves(stateAfterSecond)) {
-                            const thirdLegality: MGPFallible<DiaballikState> =
-                                DiaballikRules.get().isLegalSubMove(state, third);
-                            if (thirdLegality.isSuccess()) {
-                                moves.push(new DiaballikMove(first, MGPOptional.of(second), MGPOptional.of(third)));
-                            }
-                        }
-                    }
-                }
+        let movesInConstruction: DiaballikMoveInConstruction[] = [new DiaballikMoveInConstruction([], node.gameState)];
+        for (let i: number = 0; i < 3; i++) {
+            let nextMovesInConstruction: DiaballikMoveInConstruction[] = [];
+            for (const move of movesInConstruction) {
+                nextMovesInConstruction = nextMovesInConstruction.concat(this.addSubMove(move));
             }
+            movesInConstruction = nextMovesInConstruction;
         }
-        return moves;
+
+        return movesInConstruction.map((m: DiaballikMoveInConstruction): DiaballikMove => m.finalize(node.gameState));
     }
 
-    private getListSubMoves(state: DiaballikState): DiaballikSubMove[] {
+    private addSubMove(moveInConstruction: DiaballikMoveInConstruction): DiaballikMoveInConstruction[] {
```

Comment:
this.getChildMove ? this.getNextMoves ?
Le nommage est meh.

Reply:
ça prend un move en construction, ça lui ajoute le sous-mouvement possible suivant, d'où "addSubMove", mais ça le fait pour chaque sous mouvement. Je propose donc `addAllPossibleSubMoves`